cloudformation/slack-router-cfn-template.js

```diff
@@ -0,0 +1,94 @@
+const cf = require('@mapbox/cloudfriend')
+
+const Resources = {
+  SlackRouterApi: {
+    Type: 'AWS::ApiGatewayV2::Api',
+    Properties: {
+      Name: 'slack-router-api',
+      ProtocolType: 'HTTP',
+    },
+  },
+  SlackToLambdaIntegration: {
+    Type: 'AWS::ApiGatewayV2::Integration',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      Description: 'Integration for Slack Slash Commands',
+      IntegrationType: 'AWS_PROXY',
+      IntegrationUri: cf.join('', [
+        'arn:',
+        cf.partition,
+        ':apigateway:',
+        cf.region,
+        ':lambda:path/2015-03-31/functions/',
+        cf.getAtt('SlackRouterLambda', 'Arn'),
+        '/invocations',
+      ]),
+      PayloadFormatVersion: '2.0',
+    },
+  },
+  SlackToLambdaRoute: {
+    Type: 'AWS::ApiGatewayV2::Route',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      AuthorizationType: 'NONE',
+      RouteKey: 'POST /slack-to-lambda',
+      Target: cf.join('/', [
+        'integrations',
+        cf.ref('SlackToLambdaIntegration'),
+      ]),
+    },
+  },
+  SlackToLambdaDeploy: {
+    Type: 'AWS::ApiGatewayV2::Deployment',
+    DependsOn: 'SlackToLambdaRoute',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+    },
+  },
+  SlackRouterApiProdStage: {
+    Type: 'AWS::ApiGatewayV2::Stage',
+    Properties: {
+      ApiId: cf.ref('SlackRouterApi'),
+      AutoDeploy: 'true',
+      DeploymentId: cf.ref('SlackToLambdaDeploy'),
+      StageName: 'prod',
+    },
+  },
+  SlackRouterLambda: {
+    Type: 'AWS::Lambda::Function',
+    Properties: {
+      FunctionName: 'slack-router-lambda',
+      Handler: 'slack-router-lambda.handler',
+      Role: cf.sub(
+        'arn:aws:iam::${AWS::AccountId}:role/service-role/test-api-lambda-sns-role-6ndj69lt'
+      ),
+      Code: {
+        S3Bucket: 'lambda-andria',
+        S3Key: 'slack-router-lambda.zip',
+      },
+      Runtime: 'nodejs12.x',
+      Timeout: '30',
+    },
+  },
+  SlackRouterPermission: {
+    Type: 'AWS::Lambda::Permission',
+    Properties: {
+      Action: 'lambda:InvokeFunction',
+      FunctionName: cf.ref('SlackRouterLambda'),
+      Principal: 'apigateway.amazonaws.com',
+      SourceArn: cf.join('', [
+        'arn:',
+        cf.partition,
+        ':execute-api:',
+        cf.region,
+        ':',
+        cf.accountId,
+        ':',
+        cf.ref('SlackRouterApi'),
+        '/*/*',
+      ]),
```

Comment:
same as above - `cf.arn` will make your template much cleaner, even though `cf.join` works, technically. : )

Reply:
What ARN should we replace this with instead? I struggled with this one too since the docs confused me. It seems like API Gateway has many [different ARN formats](https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html) and looking at that list there wasn't any that matched what I needed so I had to create this long and confusing joint string instead.

Would appreciate your help on this one!

Reply:
This one is a little easier. So, if you look at the code for [`cf.arn`](https://github.com/mapbox/cloudfriend/blob/master/lib/intrinsic.js#L218-L222), the structure they follow for non S3 ARNs is:

`'arn:${AWS::Partition}:${service}:${AWS::Region}:${AWS::AccountId}:${suffix}'`

So, your `${service}` would be `execute-api` and your `${suffix}` would be `cf.join('', [cf.ref('SlackRouterApi'), '/*/*')`. You'd replace `cf.join` with `cf.arn('execute-api', cf.join('', [cf.ref('SlackRouterApi'), '/*/*' ])`

Reply:
Done c0ae532e990 Thanks for the help!